pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java

```diff
@@ -263,6 +263,29 @@ public boolean checkAndStartPublish(long producerId, long sequenceId, ByteBuf he
             }
         }
 
+        if (topic.isPersistent()) {
+            PersistentTopic pTopic = (PersistentTopic) topic;
+            if (pTopic.isDelayedDeliveryEnabled()) {
+                long maxDeliveryDelayInMs = pTopic.getDelayedDeliveryMaxDelayInMillis();
+                if (maxDeliveryDelayInMs > 0) {
+                    headersAndPayload.markReaderIndex();
+                    MessageMetadata msgMetadata = Commands.parseMessageMetadata(headersAndPayload);
+                    headersAndPayload.resetReaderIndex();
+                    if (msgMetadata.hasDeliverAtTime()
+                            && msgMetadata.getDeliverAtTime() - msgMetadata.getPublishTime() > maxDeliveryDelayInMs) {
+                        cnx.execute(() -> {
+                            cnx.getCommandSender().sendSendError(producerId, sequenceId,
+                                    ServerError.NotAllowedError,
+                                    String.format("Exceeds max allowed delivery delay of %s milliseconds",
+                                            maxDeliveryDelayInMs));
+                            cnx.completedSendOperation(false, headersAndPayload.readableBytes());
+                        });
+                        return false;
+                    }
```

Comment:
I'm not sure about it, but there's a chance that completedSendOperation should only be called if the message was accepted for sending by returning true from this method.

Reply:
The other error checks in the `checkAndStartPublish` method (right above my change) are calling `cnx.getCommandSender().sendSendError(...)` and then `cnx.completedSendOperation(...)` immediately.

For instance, the encryption error check (when encryption is required but the producer does not encrypt the message):
https://github.com/apache/pulsar/blob/32f3577a735581096d85aa961d7df45b9ae9b6f9/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java#L248-L264